staging/src/k8s.io/apiextensions-apiserver/pkg/apis/apiextensions/v1beta1/types.go

```diff
@@ -139,3 +147,84 @@ type CustomResourceDefinitionList struct {
 	// Items individual CustomResourceDefinitions
 	Items []CustomResourceDefinition `json:"items" protobuf:"bytes,2,rep,name=items"`
 }
+
+// CustomResourceValidation is a list of validation methods for CustomResources.
+type CustomResourceValidation struct {
+	// OpenAPIV3Schema is the OpenAPI v3 schema to be validated against.
+	OpenAPIV3Schema *JSONSchemaProps `json:"openAPIV3Schema,omitempty"`
+}
+
+// JSONSchemaProps is a JSON-Schema following Specification Draft 4 (http://json-schema.org/).
+type JSONSchemaProps struct {
```

Comment:
Presumably this was covered in a design doc that I didn't manage to look at, but: is it really a good idea to recreate a spec type in our API like this? Was there a debate on the pros and cons? Sorry to bring this up at this late stage. But stuff we put in our API is forever.

Reply:
What exactly is the question? Yes, there was debate. Even if we reused types from a library, it would be in our API forever. The point is that this way we can guarantee it forever. By using a library which meanwhile switched to OpenAPI v6 we maybe couldn't.